server/src/jobs.rs

```diff
@@ -0,0 +1,320 @@
+use core::fmt;
+use proto::market::FileInfoHash;
+use rand::Rng;
+use serde::Serialize;
+use std::{
+    collections::HashMap,
+    sync::Arc,
+    time::{SystemTime, UNIX_EPOCH},
+};
+use tokio::{
+    sync::{Mutex, RwLock},
+    task::JoinHandle,
+};
+
+use peernode::consumer::{
+    encode::{self, EncodedUser},
+    get_file_chunk,
+    http::GetFileResponse,
+};
+
+type AsyncJob = Arc<Mutex<Job>>;
+
+#[derive(Clone)]
+pub struct Jobs {
+    pub jobs: Arc<RwLock<HashMap<String, AsyncJob>>>,
+    pub history: Arc<RwLock<HashMap<String, HistoryEntry>>>,
+}
+
+#[derive(Debug)]
+pub struct Job {
+    job_id: String,
+    file_info_hash: FileInfoHash,
+    file_name: String,
+    file_size: u64,
+    time_queued: u64, // unix time in seconds
+    status: JobStatus,
+    accumulated_cost: u64,
+    projected_cost: u64,
+    eta: u64, // seconds
+    pub peer_id: String,
+    pub encoded_producer: EncodedUser,
+}
+
+impl Job {
+    pub fn pause(&mut self) {
+        if let JobStatus::Active(_) = self.status {
+            self.status = JobStatus::Stop;
+        }
+    }
+    pub fn terminate(&mut self) {
+        match &mut self.status {
+            JobStatus::Active(handle) => {
+                handle.abort();
+                self.status = JobStatus::Failed;
+            }
+            JobStatus::Completed => {}
+            _ => self.status = JobStatus::Failed,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum JobStatus {
+    Active(JoinHandle<()>),
+    Stop,        // wait to stop
+    Paused(u64), // chunk num of next
+    Completed,
+    Failed,
+}
+
+impl fmt::Display for JobStatus {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            JobStatus::Active(_) => write!(f, "active"),
+            JobStatus::Paused(_) => write!(f, "paused"),
+            JobStatus::Stop => write!(f, "stopping"),
+            JobStatus::Completed => write!(f, "completed"),
+            JobStatus::Failed => write!(f, "failed"),
+        }
+    }
+}
+
+pub async fn start(job: AsyncJob, token: String) {
```

Comment:
Should we move this down to impl Jobs? Once the job is finished, it needs to call self.finish_job to add it to the history.

Reply:
The task that's spawned needs to release the lock on job in order for other tasks to possibly grab the lock and pause the task (which is why it's called on `AsyncJob`). There's probably a better way to do this, or maybe `start`, `pause`, `terminate` could be moved to methods of `Jobs`.